File: LeetCode/Medium/0515-find-largest-value-in-each-tree-row/0515-find-largest-value-in-each-tree-row.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        ans = []
        q = deque([root])
        while q:
            max_val = float('-inf')
            size = len(q)
            changed = False
            for _ in range(size):
                cur = q.popleft()
                if not cur:
                    continue
                max_val = max(max_val, cur.val)
                if cur.left:
                    q.append(cur.left)
                if cur.right:
                    q.append(cur.right)
                changed = True
            
            if changed:
                ans.append(max_val)
        return ans

File: LeetCode/Medium/0515-find-largest-value-in-each-tree-row/0515-find-largest-value-in-each-tree-row.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> largestValues(TreeNode* root) {
        if (!root) return {};

        deque<TreeNode*> q;
        vector<int> ans;
        q.push_back(root);
        while(!q.empty()){
            int max_val = INT_MIN;
            int size = q.size();
            for(int i = 0; i < size; i++){
                TreeNode* cur = q.front();
                q.pop_front();
                max_val = max(max_val, cur->val);
                if (cur->left) q.push_back(cur->left); 
                if (cur->right) q.push_back(cur->right);  //
            }
            ans.push_back(max_val);
        }
        return ans;
    }
};